File: python/src/sensor.py
import time
from datetime import datetime

import uuid
import random
import dognames
import names
import argparse
from db import config
from db.client import ScyllaClient


def generate_random_pet_name():
    return dognames.male() if random.randint(1, 2)==1 else dognames.female()


def generate_random_measurement_value(sensor_type):
    if sensor_type == "T":
        return 101 + random.uniform(0, 10) - 4;
    elif sensor_type == "L":
      return 35 + random.uniform(0, 5) - 2
    return random.uniform(0, 20)


def generate_random_measurement(sensor):
    return {
        "sensor_id": sensor["sensor_id"],
        "ts": datetime.now(),
        "value": generate_random_measurement_value(sensor["type"])
    }
  
  
def create_owner():
    return {
        "owner_id": uuid.uuid4(),
        "address": "home",
        "name": names.get_full_name() # generate random person's name
    }


def create_pet(owner):
    return {
        "owner_id": owner["owner_id"],
        "pet_id": uuid.uuid4(),
        "address": "home",
        "age": random.randint(1, 20),
        "name": generate_random_pet_name(),
        "weight": random.uniform(5.0, 15.0)
    }


def create_sensors(pet):
    return [
        {
            "pet_id": pet["pet_id"],
            "sensor_id": uuid.uuid4(),
            "type": "T"
        },
        {
            "pet_id": pet["pet_id"],
            "sensor_id": uuid.uuid4(),
            "type": "L"
        }
    ]

def insert_pet_static_data(client: ScyllaClient, pet, owner, sensors):
    client.insert_data(table="carepet.pet", data_dict=pet)
    client.insert_data(table="carepet.owner", data_dict=owner)
    for sensor in sensors:
        client.insert_data(table="carepet.sensor", data_dict=sensor)
    
    
def generate_measurements(sensors, count=2, delay=1):
    measurements_list = []
    for i in range(0, int(count)):
        for sensor in sensors:
            measurement = generate_random_measurement(sensor)
            measurements_list.append(measurement)
            
            print(f"sensor # {sensor['sensor_id']} type {sensor['type']},",
                  f"new measure: {measurement['value']},",
                  f"ts: {measurement['ts']}")
        time.sleep(delay)
    return measurements_list
    
    
def init_args_parser():
    arg_parser = config.argument_parser()
    arg_parser.add_argument("-b", "--buffer-interval",
                        help="Sensors measurement interval (seconds)",
                        required=False, default=10)
    arg_parser.add_argument("-m", "--measure",
                        help="Buffer to accumulate measures (seconds)",
                        required=False, default=1)
    return vars(arg_parser.parse_args())


def main():
    # parse config arguments from command line input
    config = init_args_parser()
    
    # both arguments are defined in seconds
    measure_sec = int(config["measure"])
    buffer_interval_sec = int(config["buffer_interval"])
    
    if measure_sec > buffer_interval_sec:
        raise ValueError("`--measure` cannot be larger than `--buffer-interval`")
    
    # calculate batch size based on the `buffer_interval` argument
    batch_size = buffer_interval_sec / measure_sec
    
    # create random static data
    new_owner = create_owner()
    new_pet = create_pet(owner=new_owner)
    new_sensors = create_sensors(pet=new_pet)
    
    print("Welcome to the Pet collar simulator")
    print(f"New owner # {new_owner['owner_id']}")
    print(f"New pet # {new_pet['pet_id']}")
    for i in range(0, len(new_sensors)):
        print(f"New sensor({i}) # {new_sensors[i]['sensor_id']}")
    
    with ScyllaClient(config) as client:
        insert_pet_static_data(client, new_pet, new_owner, new_sensors)
        # infinite loop to insert random time-series data into the measurement table
        while (True):
            measurements_data = generate_measurements(new_sensors,
                                                      count=batch_size,
                                                      delay=measure_sec)
            print("Pushing data")
            client.insert_batch_data("carepet.measurement", measurements_data)
        
main()

File: csharp/Sensor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using Cassandra; // DataStax Cassandra C# driver
using Microsoft.Extensions.Logging;
using CarePet.Model;
using System.CommandLine;
using System.CommandLine.Invocation;

namespace CarePet
{
    public class Sensor
    {
        private static readonly ILogger<Sensor> LOG;
        private readonly SensorConfig _config;
        private readonly Model.Owner _owner;
        private readonly Model.Pet _pet;
        private readonly Model.Sensor[] _sensors;

        static Sensor()
        {
            using var loggerFactory = LoggerFactory.Create(builder =>
            {
                builder.AddConsole();
            });
            LOG = loggerFactory.CreateLogger<Sensor>();
        }

        public Sensor(SensorConfig config)
        {
            _config = config;
            _owner = Owner.Random();
            _pet = Pet.Random(_owner.OwnerId);
            _sensors = new CarePet.Model.Sensor[Enum.GetValues(typeof(SensorType)).Length];

            var sensorTypes = Enum.GetValues(typeof(SensorType)).Cast<SensorType>().ToArray();
            for (int i = 0; i < _sensors.Length; i++)
            {
                _sensors[i] = new CarePet.Model.Sensor(_pet.PetId, Guid.NewGuid(), SensorTypeExtensions.GetTypeCode(sensorTypes[i]));
            }
        }

        public static void Main(string[] args)
        {
            var config = SensorConfig.Parse(args);
            var client = new Sensor(config);
            client.Save();
            client.Run();
        }

        /// <summary>
        /// Initiates a connection with the configured keyspace.
        /// </summary>
        public ISession Keyspace()
        {
            var session = _config.Builder(Config.Keyspace).Build().Connect();
            session.ChangeKeyspace(Config.Keyspace);
            return session;
        }

        /// <summary>
        /// Save owner, pet, and sensors to the database.
        /// </summary>
        private void Save()
        {
            using (var session = Keyspace())
            {
                var mapper = new Mapper(session);
                LOG.LogInformation($"owner = {_owner}");
                LOG.LogInformation($"pet = {_pet}");

                mapper.Owner().Create(_owner);
                mapper.Pet().Create(_pet);

                foreach (var s in _sensors)
                {
                    LOG.LogInformation($"sensor = {s}");
                    mapper.Sensor().Create(s);
                }
            }
        }

        /// <summary>
        /// Generate random sensor data and push it to the database.
        /// </summary>
        private void Run()
        {
            using (var session = Keyspace())
            {
                var prepared = session.Prepare("INSERT INTO measurement (sensor_id, ts, value) VALUES (?, ?, ?)");
                var ms = new List<Measure>();
                var prev = DateTimeOffset.UtcNow;

                while (true)
                {
                    while ((DateTimeOffset.UtcNow - prev) < _config.BufferInterval)
                    {
                        if (!Sleep(_config.Measurement))
                            return;

                        foreach (var s in _sensors)
                        {
                            var m = ReadSensorData(s);
                            ms.Add(m);
                            LOG.LogInformation(m.ToString());
                        }
                    }

                    var elapsed = DateTimeOffset.UtcNow - prev;
                    var intervals = elapsed.Ticks / _config.BufferInterval.Ticks;
                    prev = prev.AddTicks(intervals * _config.BufferInterval.Ticks);

                    LOG.LogInformation("pushing data");

                    var batch = new BatchStatement();
                    foreach (var m in ms)
                    {
                        batch.Add(prepared.Bind(m.SensorId, m.Ts.UtcDateTime, m.Value));
                    }

                    session.Execute(batch);
                    ms.Clear();
                }
            }
        }

        private bool Sleep(TimeSpan interval)
        {
            try
            {
                Thread.Sleep(interval);
                return true;
            }
            catch (ThreadInterruptedException)
            {
                return false;
            }
        }

        private Measure ReadSensorData(CarePet.Model.Sensor s)
        {
            return new Measure(s.SensorId, DateTimeOffset.UtcNow, s.RandomData());
        }

        public class SensorConfig : Config
        {
            public TimeSpan BufferInterval { get; set; } = TimeSpan.FromHours(1);
            public TimeSpan Measurement { get; set; } = TimeSpan.FromMinutes(1);

            public static SensorConfig Parse(string[] args)
            {
                var config = new SensorConfig();

                // Base options from Config
                var hostsOption = new Option<string[]>("--hosts", "Database contact points");
                var dcOption = new Option<string>(new[] { "-dc", "--datacenter" }, "Local datacenter name");
                var usernameOption = new Option<string>(new[] { "-u", "--username" }, "Authentication username");
                var passwordOption = new Option<string>(new[] { "-p", "--password" }, "Authentication password");
                var helpOption = new Option<bool>(new[] { "-h", "--help" }, "Display help message");

                var bufferInterval = new Option<TimeSpan>(
                        "--buffer-interval",
                        "Buffer interval to accumulate measures");
                var measure = new Option<TimeSpan>(
                        "--measure",
                        "Sensors measurement interval");
                var rootCommand = new RootCommand
                {
                    hostsOption,
                    dcOption,
                    usernameOption,
                    passwordOption,
                    helpOption,
                    bufferInterval,
                    measure
                };

                rootCommand.SetHandler((InvocationContext context) =>
                {
                    config.Hosts = context.ParseResult.GetValueForOption(hostsOption);
                    config.Datacenter = context.ParseResult.GetValueForOption(dcOption);
                    config.Username = context.ParseResult.GetValueForOption(usernameOption);
                    config.Password = context.ParseResult.GetValueForOption(passwordOption);
                    config.Help = context.ParseResult.GetValueForOption(helpOption);

                    config.BufferInterval = context.ParseResult.GetValueForOption(bufferInterval);
                    config.Measurement = context.ParseResult.GetValueForOption(measure);
                });

                rootCommand.InvokeAsync(args);

                if (config.Help == true)
                {
                    rootCommand.InvokeAsync("-h");
                    Environment.Exit(1);
                }

                return config;
            }
        }
    }
}
